tests/random_assignment.py

```diff
@@ -0,0 +1,94 @@
+   (EUnaryOp('sum', EMap(EFilter(EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt())), ELambda(EVar('i').with_type(TInt()), EBinOp(EVar('i').with_type(TInt()), '>', ENum(10).with_type(TInt())).with_type(TBool()))).with_type(TBag(TInt())), ELambda(EVar('i').with_type(TInt()), ENum(1).with_type(TInt()))).with_type(TBag(TInt()))).with_type(TInt()), ENum(0).with_type(TInt()), {'l': Bag((11, 0, 10, 11))}),
+   (EBinOp(EBinOp(EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt())), '+', ESingleton(EVar('i').with_type(TInt())).with_type(TBag(TInt()))).with_type(TBag(TInt())), '-', EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TBag(TInt())), EEmptyList().with_type(TBag(TInt())), {'l': Bag(()), 'i': 0}),
+   (EStateVar(EUnaryOp('sum', EVar('l').with_type(TBag(TInt()))).with_type(TInt())).with_type(TInt()), EUnaryOp('-', ENum(1).with_type(TInt())).with_type(TInt()), {'l': Bag(())}),
+   (EBinOp(EUnaryOp('sum', EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TInt()), '+', EVar('i').with_type(TInt())).with_type(TInt()), ENum(0).with_type(TInt()), {'l': Bag((0, -2438, 2437)), 'i': 0}),
+   (EBinOp(EBinOp(EFilter(EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt())), ELambda(EVar('_var372').with_type(TInt()), EBinOp(EVar('_var372').with_type(TInt()), '>', ENum(10).with_type(TInt())).with_type(TBool()))).with_type(TBag(TInt())), '+', ECond(EBinOp(EVar('i').with_type(TInt()), '>', ENum(10).with_type(TInt())).with_type(TBool()), ESingleton(EVar('i').with_type(TInt())).with_type(TBag(TInt())), EEmptyList().with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TBag(TInt())), '-', EStateVar(EFilter(EVar('l').with_type(TBag(TInt())), ELambda(EVar('_var492').with_type(TInt()), EBinOp(EVar('_var492').with_type(TInt()), '>', ENum(10).with_type(TInt())).with_type(TBool()))).with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TBag(TInt())), EBinOp(EBinOp(EFilter(EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt())), ELambda(EVar('_var372').with_type(TInt()), EBinOp(EVar('_var372').with_type(TInt()), '>', ENum(10).with_type(TInt())).with_type(TBool()))).with_type(TBag(TInt())), '+', ECond(EBool(True).with_type(TBool()), ESingleton(EVar('i').with_type(TInt())).with_type(TBag(TInt())), EEmptyList().with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TBag(TInt())), '-', EStateVar(EFilter(EVar('l').with_type(TBag(TInt())), ELambda(EVar('_var492').with_type(TInt()), EBinOp(EVar('_var492').with_type(TInt()), '>', ENum(10).with_type(TInt())).with_type(TBool()))).with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TBag(TInt())), {'l': Bag(()), 'i': 0}),
+   (EBinOp(EUnaryOp('sum', EStateVar(EVar('l').with_type(TBag(TInt()))).with_type(TBag(TInt()))).with_type(TInt()), '+', EVar('i').with_type(TInt())).with_type(TInt()), EUnaryOp('sum', EStateVar(EVar('l').with_type(TBag(TInt()
```

Comment:
Minor nit that you may or may not care about: the context object lists all the variables that the solver should care about. It _so happens_ that the solver allows additional variables in the formula (_i.e._ the free variables in the test cases above). However, I don't know if we want to guarantee that behavior. A free variable in a formula that is not present in the context may indicate an internal bug, so we might change the solver in the future to throw an exception in this case.

A "more correct" thing would be to construct a context containing a superset of each test case's free variables. However, what you have here _does_ work today.

Reply:
(The comment above refers to the `init_solver` function; I accidentally attached it to the wrong lines!)